### PR TITLE
Added support for NVidia NVML to query GPU power and temperature

### PR DESCRIPTION
**Pull request by philipheinisch**
Adds two new command line switches -np xx and -nt xx to query GPU power and temperature using NVidia NVML, the sampling rate is controlled by the value xx in ms. For NVML to work, the NVidia CUDA toolkit needs to be installed. Temperature and power and corresponding timestamps are written to the output HDF5 file. This feature is controlled by the 'USENVML' define, so it can still be used without the CUDA toolkit. It could be made more user friendly in the future, but it works.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,15 @@ cmake_minimum_required (VERSION 3.5)
 PROJECT(toolkitICL)
 
 
+option(USENVML "Use NVML" ON)
```
Comment (ranocha) [inline CMakeLists.txt]: We must either set the default to `OFF` or disable this option explicitly for the CI tests where CUDA isn't installed.

Since OpenCL does not target CUDA exclusively, I would vote for an explicit opt-in for this option.

File: src/main.cpp [hunk 2]
```diff
@@ -13,10 +13,189 @@
...
+	return 0;
+}
 
+#else
+#include <sys/time.h>
+#endif
 
 using namespace std;
 
+#define USENVML
```
Comment (ranocha) [inline src/main.cpp]: This will result in failures if the option `USENVML` is disabled at configuration/build time.

File: src/hdf5_io.cpp [hunk 3]
```diff
@@ -313,6 +313,65 @@ template bool h5_write_buffer(char const* filename, char const* varname, cl_long
 template bool h5_write_buffer(char const* filename, char const* varname, cl_ulong const* data, size_t size);
 
 
+
+// write a buffer to an HDF5 file using compression
+template<typename TYPE>
+bool h5_append_buffer(char const* filename, char const* varname, TYPE const* data, size_t size)
```
Comment (ranocha) [inline src/hdf5_io.cpp]: How is this function different from `h5_write_buffer<TYPE>`? Aren't they exactly the same?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -17,6 +17,13 @@ set(SOURCES main.cpp hdf5_io.cpp ocl_dev_mgr.cpp ${HEADER})
 
 # Specifiy library paths for linker
 link_directories (${OpenCL_LIBRARIES} ${HDF5_HL_LIBRARIES} ${HDF5_LIBRARIES})
+IF(USENVML)
+if (MSVC)
+link_directories(${OpenCL_LIBRARIES} ${HDF5_HL_LIBRARIES} ${HDF5_LIBRARIES} ${CUDA_TOOLKIT_ROOT_DIR}/lib/x64)
+ELSE(MSVC)
+link_directories(toolkitICL ${OpenCL_LIBRARIES} ${HDF5_HL_LIBRARIES} ${HDF5_LIBRARIES} ${CUDA_TOOLKIT_ROOT_DIR}/lib/x86_64-linux-gnu/stubs/)
```
Comment (ranocha) [inline src/CMakeLists.txt]: Why did you add `toolkitICL` here?

File: src/main.cpp [hunk 5]
```diff
@@ -13,10 +13,189 @@
...
+{
+	nvmlDevice_t device;
+	nvmlDeviceGetHandleByIndex(0, &device);
+
+	while (nv_log_pwr== true) {
+		std::this_thread::sleep_for(std::chrono::milliseconds(nv_p_rate));
+
+		nvmlDeviceGetPowerUsage(device, &temp); 
+    //time(&rawtime);
+    gettimeofday(&rawtime, NULL);
```
Comment (ranocha) [inline src/main.cpp]: This part looks weird because of mixed indentation (tabs & spaces).

File: src/main.cpp [hunk 6]
```diff
@@ -13,10 +13,189 @@
...
+#include <sys/time.h>
+#endif
 
 using namespace std;
 
+#define USENVML
+
+#if defined(USENVML)
+#include <nvml.h>
+bool nv_log_pwr = false;
```
Comment (ranocha) [inline src/main.cpp:51]: I have some personal objections against global variables but that might be just because of their drawbacks in Julia :wink:

File: src/main.cpp [hunk 7]
```diff
@@ -300,7 +499,7 @@ int main(int argc, char *argv[]) {
 
   for (string const& kernel_name : kernel_list) {
     exec_time = exec_time + dev_mgr.execute_kernelNA(*(dev_mgr.getKernelbyName(0, "ocl_Kernel", kernel_name)),
-                                                     dev_mgr.get_queue(0, 0), range_start, global_range, local_range);
```
Comment (ranocha) [inline src/main.cpp]: Why did you delete the whitespace here? It is harder to read (at least for me) in this new version.

File: src/main.cpp [hunk 8]
```diff
@@ -318,7 +517,56 @@ int main(int argc, char *argv[]) {
     std::this_thread::sleep_for(timespan);
   }
 
-  cout << "Saving results... " << endl;
+ cout << "Saving results... " << endl;
+
+#if defined(USENVML)
+  nv_log_pwr = false;
+  nv_log_tmp = false;
```
Comment (ranocha) [inline src/main.cpp:528]: Could you please use a homogeneous indentation?

**Review by ranocha — CHANGES_REQUESTED** [review]

File: src/hdf5_io.cpp (hunk 3 above)
Comment (philipheinisch) [reply src/hdf5_io.cpp]: Right now it is not. The idea behind it was to be able to append data to exisiting HDF5 entires. This could be important to write live data for example the current GPU power to the HDF5 files. Right now everything gets buffered in RAM and dumped at the end. This is just a stub, because for short execution times, RAM buffering is more efficient.

File: src/CMakeLists.txt (hunk 4 above)
Comment (philipheinisch) [reply src/CMakeLists.txt]: No specific reason, must ahve been a copy&paste mistake.

File: src/main.cpp (hunk 6 above)
Comment (philipheinisch) [reply src/main.cpp:51]: It seemed like the easiest option to implement shared memory for the std::threads without having to define a specific class. I agree, that it is not the most elegant solution. Any better ideas?

File: src/main.cpp (hunk 7 above)
Comment (philipheinisch) [reply src/main.cpp]: It seems like changing between VS2017/VS2015 and VS Code has some drawbacks regarding auto indentations, whitespaces and so on.

File: src/main.cpp (hunk 5 above)
Comment (philipheinisch) [reply src/main.cpp]: It seemed like the easiest option to implement shared memory for the std::threads without having to define a specific class. I agree, that it is not the most elegant solution. Any better ideas?

Comment (philipheinisch) [reply src/main.cpp]: It seems like changing between VS2017/VS2015 and VS Code has some drawbacks regarding auto indentations, whitespaces and so on.

File: src/hdf5_io.cpp (hunk 3 above)
Comment (ranocha) [reply src/hdf5_io.cpp]: I would favor either a TODO note in the code (without implementation) or a full implementation. Right now, someone who reads the code will expect that this function can be used for the purpose of its name.

File: src/main.cpp (hunk 6 above)
Comment (ranocha) [reply src/main.cpp:51]: Local variables and a lambda/closure? You can also use global variables if you like. I won't reject your changes because of that :wink:

File: CMakeLists.txt (hunk 1 above)
Comment (philipheinisch) [reply CMakeLists.txt]: What about setting default to off, but enabling it automatically, if CUDA is detected?

Comment (ranocha) [reply CMakeLists.txt]: That sounds good.

File: src/main.cpp (hunk 6 above)
Comment (philipheinisch) [reply src/main.cpp:51]: Maybe we could use global variables for now and if all works well, wrap everything in a seperate class with correct non global variables for shared memory when we have time for cleanup. It should be done at some point anyways, otherwise the main.cpp will get overcrowded, especially if we also implement things like  Intel® Power Gadget.

File: src/hdf5_io.cpp (hunk 3 above)
Comment (philipheinisch) [reply src/hdf5_io.cpp]: For now, RAM buffering seems fine, so I will remove the stub code and add a ToDo.

**Comment by philipheinisch** [comment]
Does it correctly find the CUDA toolkit on your system now @ranocha? And how can I re-run the CI checks?

**Comment by ranocha** [comment]
Yes, it seems to find CUDA but cmake errors:
```
toolkitICL/build$ cmake ..
-- The C compiler identification is GNU 7.3.0
-- The CXX compiler identification is GNU 7.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done                                            
-- Looking for pthread.h                                                            
-- Looking for pthread.h - found                                                    
-- Looking for pthread_create                                                       
-- Looking for pthread_create - not found                                           
-- Looking for pthread_create in pthreads                                           
-- Looking for pthread_create in pthreads - not found                               
-- Looking for pthread_create in pthread                                            
-- Looking for pthread_create in pthread - found                                    
-- Found Threads: TRUE                                                              
-- Found CUDA: /usr (found version "9.1")                                           
-- Using NVML                                                                       
CMake Error at CMakeLists.txt:11 (add_compile_definitions):                         
  Unknown CMake command "add_compile_definitions".                                  
                                                                                    
                                                                                    
-- Configuring incomplete, errors occurred!                                         
```

**Comment by ranocha** [comment]
The CI tests should run automatically if you push new commits.

**Comment by philipheinisch** [comment]
What is your cmake version? Maybe add_compile_definitions (https://cmake.org/cmake/help/v3.12/command/add_compile_definitions.html) is to new of a feature to be available on most platforms? What would be an acceptable minimal cmake version? 

Is it possible, that the CI test are not working anymore due to the merge conflicts?

**Comment by ranocha** [comment]
> Is it possible, that the CI test are not working anymore due to the merge conflicts?

That would be a bit strange but I'm not sure.

**Comment by ranocha** [comment]
> What is your cmake version?

I can have a look in the afternoon. At my computer at TU BS, I'm using `cmake version 3.7.2`.

**Comment by philipheinisch** [comment]
> I can have a look in the afternoon. At my computer at TU BS, I'm using `cmake version 3.7.2`.

That is indeed very old... But could you retry at your office PC? It should work now, even with ancient CMake versions.

**Comment by ranocha** [comment]
> But could you retry at your office PC? It should work now, even with ancient CMake versions.

It can be compiled without errors (but some tests fail - I will have a look at that later).

**Comment by philipheinisch** [comment]
Ok, if it compiles correctly on different platforms, I will try and fix the reaming indentation problems and than we can probably try to merge?

**Comment by ranocha** [comment]
I still have to test it at home - I have no CUDA at my office PC.Afterwards, we can fix the remaining parts & merge conflicts. Then, CI tests will hopefully pass again.

**Comment by ranocha** [comment]
The build using CUDA is fine if I run `cmake` (version 3.10.2). However, if I use `cmake -DUSENVML=OFF ..`, the build fails because CUDA should still be linked
```
CMakeFiles/toolkitICL.dir/main.cpp.o: In function `nv_log_pwr_func()':
main.cpp:(.text+0x11d0): undefined reference to `nvmlInit_v2'
main.cpp:(.text+0x11ef): undefined reference to `nvmlDeviceGetHandleByIndex_v2'
main.cpp:(.text+0x1273): undefined reference to `nvmlDeviceGetPowerUsage'
main.cpp:(.text+0x12d0): undefined reference to `nvmlShutdown'
CMakeFiles/toolkitICL.dir/main.cpp.o: In function `nv_log_tmp_func()':
main.cpp:(.text+0x1380): undefined reference to `nvmlInit_v2'
main.cpp:(.text+0x1390): undefined reference to `nvmlDeviceGetHandleByIndex_v2'
main.cpp:(.text+0x1427): undefined reference to `nvmlDeviceGetTemperature'
main.cpp:(.text+0x148a): undefined reference to `nvmlShutdown'
collect2: error: ld returned 1 exit status
src/CMakeFiles/toolkitICL.dir/build.make:156: recipe for target 'bin/toolkitICL' failed
make[2]: *** [bin/toolkitICL] Error 1
CMakeFiles/Makefile2:85: recipe for target 'src/CMakeFiles/toolkitICL.dir/all' failed
make[1]: *** [src/CMakeFiles/toolkitICL.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Edit: But that can also be attributed to a weird use case - or a dump user :wink:

File: CMakeLists.txt [hunk 9]
```diff
@@ -3,6 +3,17 @@
...
 PROJECT(toolkitICL)
 
+ find_package(CUDA)
+ 
+IF( CUDA_FOUND)
+ option(USENVML "Use NVML" ON)
+ MESSAGE(STATUS "Using NVML")
+ #add_compile_definitions(USENVML)
+ add_definitions(-DUSENVML)
+ELSEIF(CUDA_FOUND)
```
Comment (ranocha) [inline CMakeLists.txt]: Shouldn't this be `ELSE`?

File: CMakeLists.txt [hunk 10]
```diff
@@ -3,6 +3,17 @@
 cmake_minimum_required (VERSION 3.5)
 PROJECT(toolkitICL)
 
+ find_package(CUDA)
+ 
+IF( CUDA_FOUND)
+ option(USENVML "Use NVML" ON)
```
Comment (ranocha) [inline CMakeLists.txt]: Maybe it would be better to use something like
```
IF(CUDA_FOUND)
  option(USENVML "Use NVML" ON)
ELSE()
  option(USENVML "Use NVML" OFF)
ENDIF()

IF(USENVML)
  MESSAGE(STATUS "Using NVML")
  add_definitions(-DUSENVML)
ENDIF()
```

File: CMakeLists.txt [hunk 11]
```diff
@@ -42,7 +56,7 @@ ENDIF(HDF5_FOUND)
 # Check for OpenCL
 find_package(OpenCL REQUIRED)
 IF(OpenCL_FOUND)
-  MESSAGE(STATUS "Looking for OpenCL - found at\n   ${OpenCL_INCLUDE_DIRS}\n   ${OpenCL_LIBRARY}")
+  MESSAGE(STATUS "Looking for OpenCL - found at\n   ${OpenCL_INCLUDE_DIRS}\n  ${OpenCL_LIBRARY}")
```
Comment (ranocha) [inline CMakeLists.txt:60]: Why did you remove the indentation here? Now, it looks a bit weird for me:
```
   /usr/include
  /usr/lib/x86_64-linux-gnu/libOpenCL.so
```

File: README.md [hunk 12]
```diff
@@ -9,13 +9,14 @@ to use cross platform tool to execute a set of kernels for example on compute cl
...
 the input HDF5 file. After execution, the output data, data copy- and runtime is written to the
-output HDF5 file.
+output HDF5 file. The GPU power consumption and temperature for supported Nvidia GPUs can also be logged automatically to the output HDF file.
 
 ## Usage
 To use toolkitCL the following needs to be installed:
 - OpenCL (headers and drivers)
 - HDF5
 - CMake
+- CUDA Toolkit (only for GPU power/temperature logging)
```
Comment (ranocha) [inline README.md:19]: Maybe a remark that CUDA will be detected automatically if present would be good? It doesn't need to be here but somewhere in the README would be good.

**Review by ranocha — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 10 above)
Comment (philipheinisch) [reply CMakeLists.txt]: Using Windows, this overrides 'cmake -DUSENVML=OFF' if CUDA is detected. I'm guessing this is the intention behind it?

File: CMakeLists.txt (hunk 9 above)
Comment (philipheinisch) [reply CMakeLists.txt]: Yes, it should have been...

File: CMakeLists.txt (hunk 10 above)
Comment (ranocha) [reply CMakeLists.txt]: You're right We (you?) should decide whether `USENVML` shall be an option for the user or not. If it is no option, we can use something as above. Otherwise, an additional check has to be added to the first `IF`. But I think it might be nice to turn `USENVML=OFF` if CUDA isn't found (with an additional warning, maybe?).

File: CMakeLists.txt (hunk 11 above)
Comment (philipheinisch) [reply CMakeLists.txt:60]: It wasn't intentional. VS2017 just likes to changed indentations.

**Comment by philipheinisch** [comment]
I added an explanation to the readme, is this what you imagined? I also changed the CMake file to include 
> IF(CUDA_FOUND)
>   option(USENVML "Use NVML" ON)
> ELSE()
>   option(USENVML "Use NVML" OFF)
> ENDIF()
> 
> IF(USENVML)
>   MESSAGE(STATUS "Using NVML")
>   add_definitions(-DUSENVML)
> ENDIF()
> ```

this is likely the most flexible solution.

**Review by ranocha — APPROVED** [review]

**Comment by ranocha** [comment]
Finally, the tests pass :tada: Thanks, @philipheinisch, good work as always (with a debatable indentation style :wink:).
